### PR TITLE
Update results with kruize 0.13

### PR DESCRIPTION
**Pull request by xrajesh**
### Related Issue

https://issues.redhat.com/browse/ACM-4778

### Description of changes
1. main.go
     - starting 2 go routines for create and update Objects to be listened 
 2. generateClusterId.go
     - added functions to convert data 
 3. getPerformanceProfile.go
     - updated Function to give real data with performance profile
  4. getPerformanceProfileMetrics.go
      - new function to get data using prom query
  5. performance_profile.go
      - Added new structs to objects used
  6.  prometheus.go
       - Simple log package update
   7. computeRecommendation.go
       - create Request Object and Channel
   8.  createExperiment.go
        - updated the correct values for Kurize input
   9. updateResults.go
         - new function to create UpdateResult object
   10. utils.go
         - moved the Prom client function to have a single client

### CODE REVIEW

File: pkg/kruize/getPerformanceProfile.go [hunk 1]
```diff
@@ -24,27 +24,65 @@ func NewProfileManager(profile_name string) *profileManager {
...
 	instanceProfile := *p.performanceProfile
-	for i, fv := range instanceProfile.Slo.Function_variables {
-		for j, af := range fv.Aggregation_functions {
-			af.Query = replaceTemplate(af.Query, clusterName, namespace, workloadName, containerName)
-			klog.V(9).Info("Updated aggregate function ", j)
+	var metric model.Metrics
+	var metricsList []model.Metrics
+	// var aggregationInfoList []map[string]float64
+	measurementDur = strings.TrimSuffix(measurementDur, "in")
+	for _, fv := range instanceProfile.Slo.Function_variables {
```
Comment (SherinV) [inline pkg/kruize/getPerformanceProfile.go:35]: Please post sample data for reference.

File: pkg/server/computeRecommendation.go [hunk 2]
```diff
@@ -96,7 +109,8 @@ func computeRecommendations(w http.ResponseWriter, r *http.Request) {
 
 	//createExperiment with data:
 	if err == nil {
-		LoadValues(requestName, deployments, context)
+		//LoadValues(requestName, deployments, context)
```
Comment (SherinV) [inline pkg/server/computeRecommendation.go:112]: remove comment

File: pkg/server/createExperiment.go [hunk 3]
```diff
@@ -98,19 +97,20 @@ func LoadValues(requestName string, deployments map[string][]string, context con
...
-			if err != nil {
-				klog.Infof("CreateExperiment %s profile created successfully", experimentName)
+			if err == nil {
+				klog.V(5).Infof("CreateExperiment %s profile created successfully", experimentName)
+				UpdateQueue <- requestBody
 			}
 
 		}
-
+		break
```
Comment (SherinV) [inline pkg/server/createExperiment.go:113]: Add a TODO to remove later.

File: pkg/kruize/getPerformanceProfile.go [hunk 4]
```diff
@@ -24,27 +24,65 @@ func NewProfileManager(profile_name string) *profileManager {
 	return pm
 }
 
-func (p *profileManager) GetPerformanceProfileInstance(clusterName string, namespace string,
-	workloadName string, containerName string) model.Perf_profile {
+//gets perfprof per container and returns query and name map
```
Comment (SherinV) [inline pkg/kruize/getPerformanceProfile.go:27]:
```suggestion
//gets perfprof per container from Thanos and returns array of metrics
```

File: pkg/kruize/getPerformanceProfileMetrics.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+	v1api := promv1.NewAPI(utils.PromClient)
+	res, _, err := v1api.Query(ctx, query, time.Now())
+	if err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			klog.Errorf("API query timed out: %v", err)
+		}
+		klog.Errorf("API query failed: %v", err)
+	}
+
+	vector := res.(model.Vector)
```
Comment (SherinV) [inline pkg/kruize/getPerformanceProfileMetrics.go:32]: What happens if the error is not nil? Do we return?

File: pkg/kruize/getPerformanceProfile.go [hunk 6]
```diff
@@ -24,27 +24,65 @@ func NewProfileManager(profile_name string) *profileManager {
...
+func (p *profileManager) GetPerformanceProfileInstanceMetrics(clusterName string, namespace string,
+	workloadName string, containerName string, measurementDur string) []model.Metrics {
```
Comment (jlpadilla) [inline pkg/kruize/getPerformanceProfile.go:29]: Suggestion.  Consider creating an object for each workload instance.
```go
type WorkloadInstance struct {
  clusterName    string
  namepace       string
  name           string
  containerName  string
  measurementDur  string    // Why not time.Duration ?
}

Comment (jlpadilla) [reply pkg/kruize/getPerformanceProfile.go:29]: 1. This same object can be enhanced to include the memory and cpu data.
2. The `replaceTemplate` function below becomes.
    ```go
    func (w *WorkloadInstance) replaceTemplate() {
    ...
    }
    ```

File: pkg/kruize/getPerformanceProfileMetrics.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+	for _, sample := range vector {
+		value = (float64)(sample.Value)
+	}
```
Comment (jlpadilla) [inline pkg/kruize/getPerformanceProfileMetrics.go:36]: `value` is always assigned the last value in the vector.  Is this what we want?  If yes, then you don't need the for loop, you can just access the last item.

```go
value := (float64)(vector[len(vector) -1].Value)
```

File: pkg/server/createExperiment.go [hunk 8]
```diff
@@ -50,41 +51,39 @@ type RecommendationSettings struct {
 	Threshold string `json:"threshold"`
 }
 
-func LoadValues(requestName string, deployments map[string][]string, context context.Context) {
-
-	var reqBody CreateExperiment
-	var kubeObj KubernetesObject
-	var containerDataClean []string
+func processRequest(req *Request) {
+	klog.Infof("Processing Request ..%d", len(req.Workloads))
```
Comment (jlpadilla) [inline pkg/server/createExperiment.go:55]: Could be useful to log `reg.RequestName` here.

File: pkg/server/updateResults.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+//var update_results_url = config.Cfg.KruizeURL + "/updateResults"
+
+func ProcessUpdateQueue(q chan CreateExperiment) {
+	for {
+		klog.Info("Processing update Q")
+		ce := <-q
+		updateResultRequest(&ce)
+		klog.Infof("Processed %s", ce.ExperimentName)
+	}
```
Comment (jlpadilla) [inline pkg/server/updateResults.go:36]: This loop is serializing the requests.
Use `go updateResultRequest(&ce)` to parallelize the processing.

File: pkg/helpers/generateClusterId.go [hunk 10]
```diff
@@ -27,3 +27,15 @@ func RemoveDuplicate(strSlice []string) []string {
 	}
 	return list
 }
+
+func ConvertCpuUsageToCores(cpuUsage float64) float64 {
```
Comment (jlpadilla) [inline pkg/helpers/generateClusterId.go:31]:
```suggestion
func ConvertCpuUsageToCores(millicpu float64) float64 {
```

File: pkg/helpers/generateClusterId.go [hunk 11]
```diff
@@ -27,3 +27,15 @@ func RemoveDuplicate(strSlice []string) []string {
...
+	// Convert CPU usage from millicores to cores
+	cores := cpuUsage * 1000.0
+	return cores
+}
+
+func ConvertMemoryUsageToMiB(memoryUsage float64) float64 {
```
Comment (jlpadilla) [inline pkg/helpers/generateClusterId.go:37]:
```suggestion
func ConvertBytesToMiB(bytes float64) float64 {
```

File: pkg/kruize/getPerformanceProfileMetrics.go (hunk 7 above)
Comment (xrajesh) [reply pkg/kruize/getPerformanceProfileMetrics.go:37]: @jlpadilla This did not work . So skipped it.

File: pkg/server/updateResults.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+	StartTimestamp    string                          `json:"interval_start_time"`
+	EndTimestamp      string                          `json:"interval_end_time"`
+	KubernetesObjects []model.KubernetesObjectMetrics `json:"kubernetes_objects"`
+}
+
+func init() {
+	UpdateQueue = make(chan CreateExperiment)
+}
+
+//var update_results_url = config.Cfg.KruizeURL + "/updateResults"
```
Comment (SherinV) [inline pkg/server/updateResults.go:28]: Comment to be removed?

File: pkg/helpers/generateClusterId.go [hunk 13]
```diff
@@ -27,3 +27,15 @@ func RemoveDuplicate(strSlice []string) []string {
 	}
 	return list
 }
+
+func ConvertCpuUsageToCores(millicpu float64) float64 {
+	// Convert CPU usage from millicores to cores
+	cores := millicpu * 1000.0
+	return cores
```
Comment (jlpadilla) [inline pkg/helpers/generateClusterId.go:34]:
```suggestion
	return millicpu * 1000.0
```

File: pkg/helpers/generateClusterId.go [hunk 14]
```diff
@@ -27,3 +27,15 @@ func RemoveDuplicate(strSlice []string) []string {
...
+}
+
+func ConvertMemoryUsageToMiB(bytes float64) float64 {
+	// Convert memory usage from bytes to Mebibytes (MiB)
+	miB := bytes / 1024.0 / 1024.0
+	return miB
```
Comment (jlpadilla) [inline pkg/helpers/generateClusterId.go:40]:
```suggestion
	return bytes / 1024.0 / 1024.0
```

File: pkg/server/updateResults.go (hunk 12 above)
Comment (xrajesh) [reply pkg/server/updateResults.go:28]: Yes we are using this in next PR

File: pkg/kruize/getPerformanceProfileMetrics.go (hunk 7 above)
Comment (jlpadilla) [reply pkg/kruize/getPerformanceProfileMetrics.go:37]: ok, can you add a TODO comment to revisit later?

**Review by jlpadilla — APPROVED** [review]

**Pull request merged**